cdap-app-fabric/src/test/java/co/cask/cdap/reporting/ProgramHeartBeatDatasetTest.java

```diff
@@ -0,0 +1,246 @@
+    datasetFramework = injector.getInstance(DatasetFramework.class);
+    txExecutorFactory = injector.getInstance(TransactionExecutorFactory.class);
+  }
+
+  @Test
+  public void testWritingScanningHeartBeats() throws Exception {
+    ProgramHeartbeatDataset programHeartbeatDataset = getHeartBeatStore("testBeats");
+    TransactionExecutor txnl = getTxExecutor(programHeartbeatDataset);
+
+    //  write program status "running" for program 1 starting at x
+    long startTime1 = TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis());
+    RunId runId = RunIds.generate();
+    RunRecordMeta meta =
+      getMockRunRecordMeta(NamespaceId.DEFAULT,
+                           runId, new StatusTime(ProgramRunStatus.RUNNING, startTime1), null);
+    txnl.execute(() -> {
+      programHeartbeatDataset.writeRunRecordMeta(meta, startTime1);
+    });
+
+    // write heart beat messages for 10 minutes (every minute) for this program run.
+    long endTime = startTime1 + TimeUnit.MINUTES.toSeconds(10);
+    long interval = TimeUnit.MINUTES.toSeconds(1);
+    setUpProgramHeartBeats(meta, startTime1, endTime, interval, txnl, programHeartbeatDataset);
+
+
+    //  write program status "running" for program 2 starting at x + 5
+    long startTime2 = startTime1 + TimeUnit.MINUTES.toSeconds(5);
+    RunId runId2 = RunIds.generate();
+    RunRecordMeta meta2 = getMockRunRecordMeta(NamespaceId.DEFAULT,
+                                               runId2, new StatusTime(ProgramRunStatus.RUNNING, startTime2), null);
+    txnl.execute(() -> {
+      programHeartbeatDataset.writeRunRecordMeta(meta2, startTime2);
+    });
+
+    // write heart beat messages for 5 minutes (every minute) for this program run.
+    setUpProgramHeartBeats(meta2, startTime2, endTime, interval, txnl, programHeartbeatDataset);
+
+    // program run1 runtime -> x     : x + 10
+    // program run2 runtime -> x + 5 : x + 10
+    txnl.execute(() -> {
+      byte[] startRowKey1 = Bytes.toBytes(startTime1);
+      byte[] startRowKey2 = Bytes.toBytes(startTime2);
+      byte[] endRowKey = Bytes.toBytes(endTime);
+      // end row key is exclusive, scanning from x : x + 5 should only return run1
+      Assert.assertEquals(1 , programHeartbeatDataset.scan(startRowKey1, startRowKey2).size());
+      // scanning from x : x + 10, should return both
+      Assert.assertEquals(2 , programHeartbeatDataset.scan(startRowKey1, endRowKey).size());
+      // scanning from x + 5 : x + 10, should return both
+      Assert.assertEquals(2 , programHeartbeatDataset.scan(startRowKey2, endRowKey).size());
+    });
+  }
+
+  /**
+   * setup and return mock program properties with status running, runId and time when it started running are
+   * specified by parameters
+   */
+  private RunRecordMeta getMockRunRecordMeta(NamespaceId namespaceId, RunId runId, StatusTime statusTime,
+                                             @Nullable TriggeringScheduleInfo triggeringScheduleInfo) {
```

Comment:
looks like this is always null in this test. Don't need it.

Reply:
similar comment to above, about having this return the builder and removing StatusTime.